dapps/satoshi-multiplayer/contract/sources/tournament.move

```diff
@@ -194,42 +218,110 @@ module contract::tournament {
                 hash_1: _,
                 secret_0,
                 secret_1,
-                round,
+                round: _,
             } = match;
 
-            // Check if player_0 or player_1 has secret.
-            if(option::is_none(&secret_0)) {
-                // If _0 has no secret, make _1 the winner.
+            // Make sure secrets are revealed before deciding the winner
+            assert!((option::is_some(&secret_0) && option::is_some(&secret_1)), ESecretsNotRevealed);
+
+            // Extract secrets for both players.
+            let s_0 = option::extract(&mut secret_0);
+            let s_1 = option::extract(&mut secret_1);
+
+            // Take last bytes from secrets.
+            let length_0 = vector::length(&s_0) - 1;
+            let length_1 = vector::length(&s_1) - 1;
+            let last_byte_0 = vector::borrow(&s_0, length_0);
+            let last_byte_1 = vector::borrow(&s_1, length_1);
+
+            // Get winner according to players' bytes XOR calc.
+            let result = (*last_byte_0 + *last_byte_1) % 2;
+
+            if(result == 1) {
                 vector::push_back(&mut tournament.players, player_1);
-            } else if(option::is_none(&secret_1)) {
-                // If _1 has no secret, make _0 the winner.
-                vector::push_back(&mut tournament.players, player_0);
             } else {
-                // Extract secrets for both players.
-                let s_0 = option::extract(&mut secret_0);
-                let s_1 = option::extract(&mut secret_1);
-
-                // Take last bytes from secrets.
-                let length_0 = vector::length(&s_0) - 1;
-                let length_1 = vector::length(&s_1) - 1;
-                let last_byte_0 = vector::borrow(&s_0, length_0);
-                let last_byte_1 = vector::borrow(&s_1, length_1);
-
-                // Get winner according to players' bytes XOR calc.
-                let result = (*last_byte_0 + *last_byte_1) % 2;
-
-                if(result == 1) {
-                    vector::push_back(&mut tournament.players, player_1);
-                } else {
-                    vector::push_back(&mut tournament.players, player_0);
-                };
+                vector::push_back(&mut tournament.players, player_0);
+            };
+
+            i = i - 1;
+            object::delete(id);
+        };
+
+
+        if (vector::length(&tournament.players) == 1){
+            let winner = vector::pop_back(&mut tournament.players);
+            let prize_value = balance::value(&mut tournament.prize);
```

Comment:
I think `balance::value` is redundant since you already have the prize as a balance in tournament.prize

Reply:
you need to provide the value to coin::take